include/rocksdb/write_batch.h

```diff
@@ -333,18 +336,43 @@ class WriteBatch : public WriteBatchBase {
   // Returns true if MarkRollback will be called during Iterate
   bool HasRollback() const;
 
-  // Assign timestamp to write batch.
-  // This requires that all keys (possibly from multiple column families) in
-  // the write batch have timestamps of the same format.
-  Status AssignTimestamp(const Slice& ts);
+  struct TimestampChecker final {
+    Status operator()(uint32_t /*cf*/, size_t& /*ts_sz*/) const {
+      return Status::OK();
+    }
+  };
 
+  // Experimental.
+  // Assign timestamp to write batch.
+  // This requires that all keys, if enable timestamp, (possibly from multiple
+  // column families) in the write batch have timestamps of the same format.
+  // checker: callable object to check the timestamp sizes of column families.
+  // User can call checker(uint32_t cf, size_t& ts_sz) which does the
+  // following:
+  // 1. find out the timestamp size of cf.
+  // 2. if cf's timestamp size is 0, then set ts_sz to 0 and return OK.
+  // 3. otherwise, compare ts_sz with cf's timestamp size and return
+  // Status::InvalidArgument() if different.
+  template <typename Checker = TimestampChecker>
```

Comment:
Also, the Checker aspect of this API is effectively for internal use only, because users would have to provide their own implementation of non-default template instantiation.